lib/ble.toit

```diff
@@ -257,23 +266,159 @@ class Client:
+  
+/** 
+A characteristic that can only be read by clients
+*/
+class ReadOnlyCharacteristic extends Characteristic:
+  value_/ByteArray := #[]
+
+  constructor service/Service uuid/uuid_pkg.Uuid value:  
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_READ_ONLY value
+    super service resource uuid
+    value_ = value
+
+  value -> ByteArray: return value_
+
+  value= value/ByteArray -> none:
+    ble_set_characteristics_value_ state_.resource value
+    value_ = value
+
+/**
+A characteristic that can only be written to by clients
+*/
+class WriteOnlyCharacteristic extends Characteristic:
+  constructor service/Service uuid/uuid_pkg.Uuid:
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_WRITE_ONLY null
+    super service resource uuid
+
+  value -> ByteArray?:
+    state_.wait_for_state BLE_WAIT_RECV 
+    data := ble_get_characteristics_value_ state_.resource
+    state_.clear_state BLE_WAIT_RECV
+    return data;
+
+/**
+A characteristic that allows both read and write by the client
+*/
+class ReadWriteCharacteristic extends Characteristic:
+  constructor service/Service uuid/uuid_pkg.Uuid value:  
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_READ_WRITE value
+    super service resource uuid
+
+  value -> ByteArray?:
+    state_.wait_for_state BLE_WAIT_RECV 
+    data := ble_get_characteristics_value_ state_.resource
+    state_.clear_state BLE_WAIT_RECV
+    return data;
+
+  value= value/ByteArray -> none:
+    ble_set_characteristics_value_ state_.resource value
+
+
+/** 
+A characteristic that the client can subscribe to changes on
+*/
+class NotificationCharacteristic extends Characteristic:
+  constructor service/Service uuid/uuid_pkg.Uuid:  
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_NOTIFICATION null
+    super service resource uuid
+
+  value= value/ByteArray -> none:
+    ble_notify_characteristics_value_ state_.resource value
+
 /**
 The local BLE device.
 
 After construction, the BLE stack is initialized and ready to use.
+
+If services is not empty, sets up the services for the advertiser. 
 */
 class Device:
   resource_group_ := ?
   resource_state_/monitor.ResourceState_? := null
 
-  constructor.default:
-    resource_group_ = ble_init_
+  server_config/ServerConfig?
+
+  constructor.default .server_config/ServerConfig?=null:
+    server_config_resource_group := server_config!=null?server_config.resource_group_:null
+    resource_group_ = ble_init_ server_config_resource_group
     add_finalizer this:: this.close
     try:
       gap := ble_gap_ resource_group_
       resource_state := monitor.ResourceState_ resource_group_ gap
-      resource_state.wait_for_state STARTED_EVENT_
+      state := resource_state.wait_for_state STARTED_EVENT_
       resource_state_ = resource_state
+
```

Comment:
Intentional edit?

Reply:
Yes. At some point I had a lot of extra code here to try catch some state that happened not to be important. So git kind of never reverted back to the original.